DAOTracker Support

dOrgJelli opened the pull request.

orenyodfat commented on ops/mappings.json:
```diff
@@ -610,6 +610,13 @@
+        "contractName": "DAOTracker",
+        "dao": "base",
+        "mapping": "DAOTracker",
+        "arcVersion": "0.0.1-rc.29"
```
it should be 0.0.1-rc.31

dOrgJelli replied:
The contract was introduced in rc.29, and the address & ABI haven't changed since, so I think it should be rc.29 no?

orenyodfat replied:
the private network (ganache) mapping should include all latest version contracts.
(need to set the number to 18 here https://github.com/daostack/subgraph/blob/master/test/sanity.spec.ts#L15 )

dOrgJelli replied:
Updated

orenyodfat commented on src/mappings/DAOTracker/datasource.yaml:
```diff
@@ -0,0 +1,20 @@
+  - mapping: DAOToken
+    version: 0.0.1-rc.31
+  - mapping: Reputation
+    version: 0.0.1-rc.31
```
where the schemes templates (e.g GenericScheme) is added ?

dOrgJelli replied:
The scheme contracts do not need to be templates because their addresses are known at startup. We cannot support dynamically tracking schemes because we don't currently have a way of determining the "Scheme Type" when a new scheme is added to the controller. We will once Arc Factories are a thing though.

orenyodfat replied:
well, you can determining the `Scheme Type` via contractInfo , via its address

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+  ResetDAO as ResetDAOEntity,
+} from '../../types/schema';
+import {
+  Avatar_0_0_1_rc_31,
```
what is this postfix is for ? Avatar_0_0_1_rc_31 ?

dOrgJelli replied:
These names are currently auto generated [here](https://github.com/dOrgTech/subgraph/blob/ab7135abeb86ad27997b1d92acb7811ba8fc6ef9/ops/generate-subgraph.js#L137) based on the version stated [here](https://github.com/dOrgTech/subgraph/blob/ab7135abeb86ad27997b1d92acb7811ba8fc6ef9/src/mappings/DAOTracker/datasource.yaml#L7).

The reason is because datasource templates cannot have the same name, and we need a way of supporting multiple versions of the ABI.

orenyodfat replied:
is that can be avoided somehow by getting the template from a pre defined mapping or something ?

orenyodfat replied:
see https://github.com/daostack/subgraph/pull/380/files#r342098970 for a reference

dOrgJelli replied:
Responded to your comment linked above, will look into this.

dOrgJelli replied:
Looked into it, and responded to your comment here: https://github.com/daostack/subgraph/pull/380/files#r342098970

orenyodfat commented on ops/generate-subgraph.js:
```diff
@@ -130,6 +132,12 @@ function combineFragments(fragments, isTemplate, addresses, missingAddresses) {
 
+    if (isTemplate) {
+      // convert name to be version specific
+      const classVersion = version.replace(/\.|-/g, '_');
```
how about generate a pre compiled mapping from for all specific version template ?
this will make the code more readable and maintainable.

dOrgJelli replied:
I'm not sure what you mean by this.

orenyodfat replied:
see https://github.com/daostack/subgraph/pull/380/files#r342098970

dOrgJelli replied:
Will look into this.

dOrgJelli replied:
Looked into it and responded to your comment here, it's impossible to do this: https://github.com/daostack/subgraph/pull/380/files#r342098970

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+import { equalStrings } from '../../utils';
+
+export function getDAOTrackerContract(address: Address): DAOTrackerContract {
+  let daoTracker = DAOTrackerContract.load(address.toHex()) as DAOTrackerContract;
```
does the DAOTrackerContract is used by any anyone ? is that really needed ?

dOrgJelli replied:
I think it's necessary for administrative debugging / checking. Also in the UI if a user tries to view a DAO that's been blacklisted, we'd want to be able to show the reason as to why it's been blacklisted from the UI.

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+  let _reputation = event.params._reputation;
+  let _daoToken = event.params._daoToken;
+
+  /* TODO: uncomment when this issue is resolved https://github.com/graphprotocol/graph-node/issues/1333
```
this issue is closed

orenyodfat replied:
I guess we need to have a pre compiled Daos black list.
the subgraph will need to be re indexed to exclude the blacklist dao .(at least on the first phase of the implementation ).

dOrgJelli replied:
Yes we'll need to have the precompiled list instead of checking this at "index time" since we can't view the "latest" version of the registry :/

dOrgJelli replied:
I'll remove this from the code. This also means that we don't need an on-chain mapping of address=>boolean, and we can instead just use the emitted events.

dOrgJelli replied:
Should we remove this from the contract and re-deploy?

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+  */
+
+  // If the avatar already exists, early out
+  if (store.get('AvatarContract', _avatar.toHex()) != null) {
```
if (AvatarContract.load(_avatar.toHex()) != null) {
..

dOrgJelli replied:
Fixed

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+    Controller_0_0_1_rc_31.create(_controller);
+  }
+
+  // Note, no additional work is needed here because...
```
how about tracking the dao schemes ( e.g GenericScheme) ?

dOrgJelli replied:
We have no way of determining what the type of the scheme is when it's added to the controller.

orenyodfat replied:
can be read from the contractInfo

dOrgJelli replied:
Yes but only for schemes we know the address of on startup.

Tracking non-universal schemes the subgraph already knows the address of at startup is handled.

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+  // Ensure the DAOTrackerContract has been added to the store
+  getDAOTrackerContract(event.address);
+
+  let _avatar = event.params._avatar;
```
why local functions vars starts with _ ?

dOrgJelli replied:
Whoops, apologies, fixing

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+    }
+  }
+
+  store.set('BlacklistedDAO', blacklistedDAO.id, blacklistedDAO);
```
BlacklistedDAO.save()

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,127 @@
+  let _explanationHash = event.params._explanationHash;
+
+  // Add the BlacklistedDAO to the store
+  let blacklistedDAO = new BlacklistedDAO(_avatar.toHex());
```
where does the blacklist is check ? and how it is implemented ?
I guess that on first phase , we should have a pre compiled black list daos .
subgraph will need to be reindexed to ignore the blacklisted dao

dOrgJelli replied:
See comment above. Yes we'll need to have a precompiled list and reindex to reflect that, and Alchemy could query the blacklist to see if it wants to show a DAO and also be able to show the reason why the DAO was blacklisted.

orenyodfat commented on src/mappings/Reputation/datasource.yaml:
```diff
@@ -10,3 +10,5 @@ eventHandlers:
     handler: handleMint
   - event: Burn(indexed address,uint256)
     handler: handleBurn
+  - event: OwnershipTransferred(indexed address,indexed address)
```
why is that needed ?

dOrgJelli replied:
I thought it made sense since DAOToken and Avatar track this event and instantate the entity in the store based off of it. Also I htink there could be cases where a new DAO is deployed without any initial REP holders, and we would still want the ReputationContract entity to exist in the store for completeness sake.

A good example of this is when you have a token holder DAO that relies on token holders redeeming reputation. When the DAO is first deployed there may be zero REP holders.

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,121 @@
+
+  // Tell the subgraph to start indexing events from the:
+  // Avatar, Controller, DAOToken, and Reputation contracts
+  Avatar_0_0_1_rc_31.create(avatar);
```
how about : 
- create a template_create module which hold all templates history.
then the code should looks like:  
```
let contractInfo = ContractInfo.load(avatar.toHex());
  if (contractInfo != null) {
     tempate_create("avatar",contractInfo.version);
  }
```

dOrgJelli replied:
Oooo I really like this approach. Will look into how to implement this.

dOrgJelli replied:
Okay so this actually doesn't make sense @orenyodfat, since Avatar is a newly deployed Avatar that we don't know the address of at build-time. This means that it wouldn't be contained within ContractInfo.

We cannot know the ABI version of these newly added / tracked contracts for Avatar, Controller, DAOToken, and Reputation since we don't have factories that're spawning them to go off of.

orenyodfat replied:
so we cannot be sure that the avatar is from v 0_0_1_rc_31 either. right ?
how about `TrackDAO` event will give us clue re arc version ?

- DAOCreator will be index and added to contractInfo.
- DAOTracker track will get arc version as a param : 
 `function track(Avatar _avatar, Controller _controller, string _arcVersion)`
- TrackDAO will emit the msg.sender and _arcVersion  along with all other addresses : 
`    event TrackDAO(address indexed _avatar, address _controller, address _reputation, address _daoToken, address _sender, string _arcVersion);
- if  _sender is DAOCreator (we can check that via contractInfo) so the avatar version is the DAOCreator version. - in this case the _arcVersion is ignored.
-  if  _sender is not DAOCreator the avatar version is _arcVersion.

dOrgJelli replied:
https://github.com/daostack/arc/pull/683

dOrgJelli replied:
https://github.com/daostack/migration/pull/219

dOrgJelli replied:
Should the DAOTracker mapping support both the old & new TrackDAO events?

dOrgJelli commented:
@orenyodfat tests are passing

dOrgJelli commented:
@orenyodfat the new DAOTracker contract that contains the arcVersion is integrated. Here's an overview of the new changes:
* Added the `ops/templates.json` file to keep track of what ABI versions should be used with what mappings.
* Updated `generate-contractsInfos.js` to...
`setTemplateInfo('Avatar', '0.0.1-rc.16', 'Avatar_0_0_1_rc_16');`
for each template templatable ABI. `setTemplateInfo(...)` is defined in `src/utils.ts`.
* Updated `generate-subgraph.js` to add all possible datasource templates to the subgraph manifest.
* Updated `DAOTracker/mapping.ts` to use the new way of instantiating templates.

Open question:
* In `DAOTracker/mapping.ts`, at line # 70, how do you think we should handle the error case where a template isn't found? Should we just fail silently and return from the handler?

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,146 @@
+  let daoToken = event.params._daoToken;
+  let sender = event.params._sender;
+  let arcVersion = event.params._arcVersion;
+
```
these local vars are redundant here.

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,146 @@
+                        reputationTemplate == null ||
+                        daoTokenTemplate == null;
+
+  let universalController = UControllerOrganization.load(controller.toHex()) != null;
```
`!= null` ?

dOrgJelli replied:
If loading the UControllerOrganization != null, it means we're using the universalController. If we're not using the universal controller, we want to (1) check to see if we're missing the template for the controller version, and (2) create the template if it exists.

orenyodfat replied:
oh, the syntax is exceptional .
`let universalController = (UControllerOrganization.load(controller.toHex()) != null);` this might be more readable ..

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,146 @@
+  }
+
+  // Note, no additional work is needed here because...
+  // * ControllerOrganization is added to the store by the 'RegisterScheme' event
```
does the tests cover that ?
indexing a Dao with its schemes ?
how about none universal schemes ? like genesisProrocol ?

dOrgJelli replied:
Yes the test currently adds a ContributionReward scheme and ensures it shows up. Non-universal schemes work as long as the subgraph knows about its address and type in the ContractInfos (at startup).

orenyodfat replied:
we need to find a way to index NonUni schemes as well without the subgraph knows its address and type in compile time.
we might need additional event for that from daotracker @dOrgJelli

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,146 @@
+      arcVersion = daoCreatorInfo.version;
+    }
+  }
+
```
need to confirm (somehow) that creating a template for wrong contract does not crash the subgraph .
this can happened only if the event does not come from daocreator.
if we cannot confirm ,so lets consider  limit this feature only for daocreator.

dOrgJelli replied:
The subgraph does not (and cannot) check to make sure the contract bytecode matches the bytecode in the ABI file. This means that it only watches for events ABIs from a given contract address, and if they do not emit from that address the subgraph will not do anything.

So in the case you're describing, the subgraph would not crash when it creates the template, it would rather just do nothing from there on out. The only case where this could happen is someone wrote custom contracts that emit the events we're looking for, and the subgraph responds to those events from these newly created contracts. The plan here, is that if DAOstack detects a "DAO" is doing this, it would blacklist it.

dOrgJelli replied:
Writing a test to verify the subgraph doesn't crash.

orenyodfat replied:
as far as I see the only way to mitigate that and stay on the safe side is to track this event only if it come via daocreator. otherwise ignore it.

orenyodfat commented:
> @orenyodfat the new DAOTracker contract that contains the arcVersion is integrated. Here's an overview of the new changes:
> 
> * Added the `ops/templates.json` file to keep track of what ABI versions should be used with what mappings.
> * Updated `generate-contractsInfos.js` to...
>   `setTemplateInfo('ContractName', 'ArcVersion', 'UniqueTemplateName');`
>   for each templatable ABI. `setTemplateInfo(...)` is defined in `src/utils.ts`.
>   ex: `setTemplateInfo('Avatar', '0.0.1-rc.16', 'Avatar_0_0_1_rc_16');`
> * Updated `generate-subgraph.js` to add all possible datasource templates to the subgraph manifest.
> * Updated `DAOTracker/mapping.ts` to use the new way of instantiating templates.
> 
> Open question:
> 
> * In `DAOTracker/mapping.ts`, at line # 70, how do you think we should handle the error case where a template isn't found? Should we just fail silently and return from the handler? This essentially means (assuming correct on-chain data) the subgraph is out of date and is missing ABIs + mappings. I propose we add a new entity to the store when this happens, and our automation looks for this entity existing to alerts the team letting them know what's going wrong.

this can happened only when the event does not come from daocreator. right ? lets consider supporting this feature only if it come from daocreator.

dOrgJelli commented:
@orenyodfat yes this only happens when it comes not from the DAOcreator. I agree that we could limit this to just the DAOcreator for now, and the work done here would still be reusable for when the factories are created.

orenyodfat commented on src/mappings/DAOTracker/mapping.ts:
```diff
@@ -0,0 +1,146 @@
+  return daoTracker;
+}
+
+export function handleTrackDAO(event: TrackDAO): void {
```
check for precompiled blacklist daos

dOrgJelli commented:
@orenyodfat The only DAOs that will be tracked are ones created by the DaoCreator. Additionally the embedded blacklist is implemented in the `ops/blacklist.json` file.
